client/material/popover.jsx

```diff
@@ -231,19 +237,27 @@ export default class Popover extends React.Component {
     if (keepInWindow && this._ref.current) {
       const refRect = this._ref.current.getBoundingClientRect()
 
-      for (const pos of ['bottom', 'top']) {
-        if (popoverPosition[pos] !== undefined) {
-          if (popoverPosition[pos] <= safeZoneVertical) {
-            popoverPosition[pos] = safeZoneVertical
-          } else if (popoverPosition[pos] + refRect.height >= clientHeight) {
-            popoverPosition[pos] = clientHeight - refRect.height - safeZoneVertical
-          }
+      if (popoverPosition.top !== undefined) {
+        if (popoverPosition.top <= safeZoneVerticalTop || refRect.height >= clientHeight) {
+          popoverPosition.top = safeZoneVerticalTop
+          popoverPosition.height = clientHeight - safeZoneVerticalTop - safeZoneVerticalBottom
```

Comment:
I think the actual change you need to make is to give `Container` in here a max-height of `100%`, the lack of a height on there seems to be the real source of the issues here.

An easier way to accomplish the safe zones alongside that might be through padding on that element? Unsure, just a thought I had while looking at it.

It'd probably also be nice to allow people to pass a className that gets applied to `Container` (such that this is styleable and if we want a shorter max-height (like we might for notifications) that is possible in one-offs

Reply:
Yeah, you're right; `max-height` is enough to constrain the popover fit inside the window, instead of these shenanigans that I did ^^. It's not exactly the same, since we can't really specify safe zone on top and bottom separately, but it's good enough.

However, scrollbars don't work with `max-height` alone it seems, so `height` needs to be explicitly set as well. So I also had to set `height` in javascript once the popover's height is actually calculated based on its contents (and limited to `max-height`).

Reply:
I think scrollbars will work fine without setting height in any pixel value, once you've set max-height on `Container`, you can set `height` on one of the children to 100% (and/or set `overflow`, one of those two things should make the layout constrain properly even with nested flexes). At no point should setting the height via JS be necessary (and doing so is quite weird since this component is also asking the component for its height post-layout)

Reply:
As far as I know, setting `height` to 100% only works if some element above it also has a fixed `height` set; it doesn't work with `max-height` alone, even though I'm not really sure why. I've tried doing what you suggested, but I just couldn't get it to work. I'd be happy to be wrong and learn something new about CSS :d